File: myExpenses/src/main/java/org/totschnig/myexpenses/model/ContribFeature.java
/*   This file is part of My Expenses.
 *   My Expenses is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   My Expenses is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with My Expenses.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.totschnig.myexpenses.model;

import android.annotation.SuppressLint;
import android.content.Context;

import org.totschnig.myexpenses.BuildConfig;
import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.util.licence.LicenceHandler;
import org.totschnig.myexpenses.util.licence.LicenceStatus;

import java.util.Date;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.core.text.HtmlCompat;

import static org.totschnig.myexpenses.util.licence.LicenceStatus.CONTRIB;
import static org.totschnig.myexpenses.util.licence.LicenceStatus.EXTENDED;
import static org.totschnig.myexpenses.util.licence.LicenceStatus.PROFESSIONAL;

//TODO separate enum definition from handler
public enum ContribFeature {
  ACCOUNTS_UNLIMITED(TrialMode.NONE) {
    @Override
    public String buildUsageLimitString(Context context) {
      String currentLicence = getCurrentLicence(context);
      return context.getString(R.string.dialog_contrib_usage_limit_accounts, FREE_ACCOUNTS, currentLicence);
    }
  },
  PLANS_UNLIMITED(TrialMode.NONE) {
    @Override
    public String buildUsageLimitString(Context context) {
      String currentLicence = getCurrentLicence(context);
      return context.getString(R.string.dialog_contrib_usage_limit_plans, FREE_PLANS, currentLicence);
    }
  },
  SPLIT_TRANSACTION,
  DISTRIBUTION,
  PRINT,
  AD_FREE(TrialMode.NONE),
  CSV_IMPORT(TrialMode.NUMBER_OF_TIMES, EXTENDED),
  SYNCHRONIZATION(TrialMode.DURATION, EXTENDED),
  SPLIT_TEMPLATE(TrialMode.NONE, PROFESSIONAL) {
    @Override
    public String buildUsageLimitString(Context context) {
      String currentLicence = getCurrentLicence(context);
      return context.getString(R.string.dialog_contrib_usage_limit_split_templates, currentLicence);
    }
  },
  PRO_SUPPORT(TrialMode.NONE, PROFESSIONAL),
  ROADMAP_VOTING(TrialMode.NONE, PROFESSIONAL),
  HISTORY(TrialMode.NUMBER_OF_TIMES, PROFESSIONAL),
  BUDGET(TrialMode.DURATION, PROFESSIONAL),
  OCR(TrialMode.DURATION, PROFESSIONAL),
  WEB_UI(TrialMode.DURATION, PROFESSIONAL),
  CATEGORY_TREE(TrialMode.UNLIMITED, PROFESSIONAL);

  private enum TrialMode {NONE, NUMBER_OF_TIMES, DURATION, UNLIMITED}

  ContribFeature() {
    this(TrialMode.NUMBER_OF_TIMES);
  }

  ContribFeature(TrialMode trialMode) {
    this(trialMode, CONTRIB);
  }

  ContribFeature(TrialMode trialMode, LicenceStatus licenceStatus) {
    this.trialMode = trialMode;
    this.licenceStatus = licenceStatus;
  }


  public static final int FREE_PLANS = 3;
  public static final int FREE_ACCOUNTS = 5;
  public static final int FREE_SPLIT_TEMPLATES = 1;
  private final int TRIAL_DURATION_DAYS = 60;

  private final TrialMode trialMode;
  private final LicenceStatus licenceStatus;
  /**
   * how many times contrib features can be used for free
   */
  public static final int USAGES_LIMIT = BuildConfig.DEBUG ? Integer.MAX_VALUE : 10;

  @NonNull
  public String toString() {
    return name().toLowerCase(Locale.US);
  }

  private int getUsages(PrefHandler prefHandler) {
    return prefHandler.getInt(getPrefKey(), 0);
  }

  /**
   * @return number of remaining usages (> 0, if usage still possible, <= 0 if not)
   */
  public int recordUsage(PrefHandler prefHandler, LicenceHandler licenceHandler) {
    if (!licenceHandler.hasAccessTo(this)) {
      if (trialMode == TrialMode.NUMBER_OF_TIMES) {
        int usages = getUsages(prefHandler) + 1;
        prefHandler.putInt(getPrefKey(), usages);
        return USAGES_LIMIT - usages;
      } else if (trialMode == TrialMode.DURATION) {
        long now = System.currentTimeMillis();
        if (getStartOfTrial(0L, prefHandler) == 0L) {
          prefHandler.putLong(getPrefKey(), now);
        }
        if (getEndOfTrial(now, prefHandler) < now) {
          return 0;
        }
      }
    }
    return USAGES_LIMIT;
  }


  private long getStartOfTrial(long defaultValue, PrefHandler prefHandler) {
    return prefHandler.getLong(getPrefKey(), defaultValue);
  }

  private long getEndOfTrial(long defaultValue, PrefHandler prefHandler) {
    long trialDurationMillis = (TRIAL_DURATION_DAYS * 24 * 60) * 60 * 1000L;
    return getStartOfTrial(defaultValue, prefHandler) + trialDurationMillis;
  }

  private String getPrefKey() {
    final String format = trialMode == TrialMode.DURATION ? "FEATURE_%s_FIRST_USAGE" : "FEATURE_USAGES_%s";
    return String.format(Locale.ROOT, format, name());
  }

  public int usagesLeft(PrefHandler prefHandler) {
    switch (trialMode) {
      case NUMBER_OF_TIMES:
        return USAGES_LIMIT - getUsages(prefHandler);
      case DURATION:
        long now = System.currentTimeMillis();
        return getEndOfTrial(now, prefHandler) < now ? 0 : 1;
      case UNLIMITED:
        return Integer.MAX_VALUE;
      default:
        return 0;
    }
  }

  public String buildRequiresString(Context ctx) {
    return ctx.getString(R.string.contrib_key_requires, ctx.getString(getLicenceStatus().getResId()));
  }

  public int getLabelResIdOrThrow(Context ctx) {
    int resId = getLabelResId(ctx);
    if (resId == 0) {
      throw new IllegalStateException("Label not defined for ContribFeature " + this);
    }
    return resId;
  }

  @VisibleForTesting
  public int getLabelResId(Context ctx) {
    String name = "contrib_feature_" + this + "_label";
    return ctx.getResources().getIdentifier(name, "string", ctx.getPackageName());
  }

  public String getLimitReachedWarning(Context ctx) {
    return ctx.getString(R.string.warning_trial_limit_reached, ctx.getString(getLabelResIdOrThrow(ctx)));
  }

  public CharSequence buildFullInfoString(Context ctx) {
    return HtmlCompat.fromHtml(ctx.getString(R.string.dialog_contrib_premium_feature,
        "<i>" + ctx.getString(getLabelResIdOrThrow(ctx)) + "</i>",
        ctx.getString(getLicenceStatus().getResId())) + " " +
        buildUsageLimitString(ctx), HtmlCompat.FROM_HTML_MODE_LEGACY);
  }

  @SuppressLint("DefaultLocale")
  @Nullable
  public CharSequence buildUsagesLeftString(Context ctx, PrefHandler prefHandler) {
    if (trialMode == TrialMode.NUMBER_OF_TIMES) {
      int usagesLeft = usagesLeft(prefHandler);
      return ctx.getText(R.string.dialog_contrib_usage_count) + " : " +
          String.format("%d/%d", usagesLeft, USAGES_LIMIT);
    } else if (trialMode == TrialMode.DURATION) {
      long now = System.currentTimeMillis();
      long endOfTrial = getEndOfTrial(now, prefHandler);
      if (endOfTrial < now) {
        return getLimitReachedWarning(ctx);
      } else {
        return ctx.getString(R.string.warning_limited_trial, ctx.getString(getLabelResIdOrThrow(ctx)),
            Utils.getDateFormatSafe(ctx).format(new Date(endOfTrial)));
      }
    } else return null;
  }

  public String buildUsageLimitString(Context context) {
    String currentLicence = getCurrentLicence(context);
    switch (trialMode) {
      case NUMBER_OF_TIMES:
        return context.getString(R.string.dialog_contrib_usage_limit, USAGES_LIMIT, currentLicence);
      case DURATION:
        return context.getString(R.string.dialog_contrib_usage_limit_synchronization, TRIAL_DURATION_DAYS, currentLicence);
      case UNLIMITED:
        return context.getString(R.string.dialog_contrib_usage_limit_with_dialog, currentLicence);
      default:
        return "";
    }
  }

  @NonNull
  protected String getCurrentLicence(Context context) {
    LicenceStatus licenceStatus = MyApplication.getInstance().getLicenceHandler().getLicenceStatus();
    return context.getString(licenceStatus == null ? R.string.licence_status_free : licenceStatus.getResId());
  }

  public CharSequence buildRemoveLimitation(Context ctx, boolean asHTML) {
    int resId = R.string.dialog_contrib_reminder_remove_limitation;
    return asHTML ? Utils.getTextWithAppName(ctx, resId) :
        ctx.getString(resId).replace(String.format("{%s}", Utils.PLACEHOLDER_APP_NAME), ctx.getString(R.string.app_name));
  }

  public boolean isExtended() {
    return getLicenceStatus() == EXTENDED;
  }

  public boolean isProfessional() {
    return getLicenceStatus() == PROFESSIONAL;
  }

  public int trialButton() {
    return (trialMode == TrialMode.UNLIMITED) ? R.string.dialog_remind_later : R.string.dialog_contrib_no;
  }

  public LicenceStatus getLicenceStatus() {
    return licenceStatus;
  }

}

File: myExpenses/src/main/java/org/totschnig/myexpenses/model/ContribFeature.kt
/*   This file is part of My Expenses.
 *   My Expenses is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   My Expenses is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with My Expenses.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.totschnig.myexpenses.model

import android.content.Context
import androidx.core.text.HtmlCompat
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.util.Utils
import org.totschnig.myexpenses.util.distrib.DistributionHelper.isPlay
import org.totschnig.myexpenses.util.licence.LicenceHandler
import org.totschnig.myexpenses.util.licence.LicenceHandler.Companion.TRIAL_DURATION_DAYS
import org.totschnig.myexpenses.util.licence.LicenceStatus
import java.util.Date

class ContribFeatureNotAvailableException(message: String): Exception(message)

enum class ContribFeature(
    val trialMode: TrialMode = TrialMode.DURATION,
    val licenceStatus: LicenceStatus = LicenceStatus.CONTRIB
) {
    ACCOUNTS_UNLIMITED(TrialMode.NONE),
    PLANS_UNLIMITED(TrialMode.NONE),
    SPLIT_TRANSACTION,
    DISTRIBUTION,
    PRINT,
    AD_FREE(TrialMode.NONE),
    CSV_IMPORT(licenceStatus = LicenceStatus.EXTENDED),
    SYNCHRONIZATION(licenceStatus = LicenceStatus.EXTENDED),
    SPLIT_TEMPLATE(TrialMode.NONE, LicenceStatus.PROFESSIONAL),
    PRO_SUPPORT(TrialMode.NONE, LicenceStatus.PROFESSIONAL),
    ROADMAP_VOTING(
        TrialMode.NONE,
        LicenceStatus.PROFESSIONAL
    ),
    HISTORY(licenceStatus = LicenceStatus.PROFESSIONAL),
    BUDGET(licenceStatus = LicenceStatus.PROFESSIONAL),
    OCR(licenceStatus = LicenceStatus.PROFESSIONAL),
    WEB_UI(licenceStatus = LicenceStatus.PROFESSIONAL),
    CATEGORY_TREE(TrialMode.UNLIMITED, LicenceStatus.PROFESSIONAL),
    BANKING(licenceStatus = LicenceStatus.PROFESSIONAL),
    AUTOMATIC_FX_DOWNLOAD(licenceStatus = LicenceStatus.PROFESSIONAL)
    ;

    enum class TrialMode {
        NONE, DURATION, UNLIMITED
    }

    override fun toString(): String {
        return name.lowercase()
    }

    val prefKey: String
        get() = "FEATURE_${name}_FIRST_USAGE"

    fun buildRequiresString(ctx: Context) =
        ctx.getString(R.string.contrib_key_requires, ctx.getString(licenceStatus.resId))

    val labelResId
        get() = when (this) {
            ACCOUNTS_UNLIMITED -> R.string.contrib_feature_accounts_unlimited_label
            PLANS_UNLIMITED -> R.string.contrib_feature_plans_unlimited_label
            SPLIT_TRANSACTION -> R.string.contrib_feature_split_transaction_label
            DISTRIBUTION -> R.string.contrib_feature_distribution_label
            PRINT -> R.string.contrib_feature_print_label
            AD_FREE -> R.string.contrib_feature_ad_free_label
            CSV_IMPORT -> R.string.contrib_feature_csv_import_label
            SYNCHRONIZATION -> R.string.contrib_feature_synchronization_label
            SPLIT_TEMPLATE -> R.string.contrib_feature_split_template_label
            PRO_SUPPORT -> R.string.contrib_feature_pro_support_label
            ROADMAP_VOTING -> R.string.contrib_feature_roadmap_voting_label
            HISTORY -> R.string.contrib_feature_history_label
            BUDGET -> R.string.contrib_feature_budget_label
            OCR -> R.string.contrib_feature_ocr_label
            WEB_UI -> R.string.contrib_feature_web_ui_label
            CATEGORY_TREE -> R.string.contrib_feature_category_tree_label
            BANKING -> R.string.title_fints
            ContribFeature.AUTOMATIC_FX_DOWNLOAD -> R.string.enable_automatic_daily_exchange_rate_download
        }

    fun getLimitReachedWarning(ctx: Context): String {
        return ctx.getString(
            R.string.warning_trial_limit_reached,
            ctx.getString(labelResId)
        )
    }

    open fun buildTrialString(
        context: Context,
        licenceHandler: LicenceHandler = context.injector.licenceHandler()
    ): CharSequence? {
        val currentLicence = getCurrentLicence(context, licenceHandler)

        return when (trialMode) {
            TrialMode.NONE -> when (this) {
                ACCOUNTS_UNLIMITED -> context.getString(
                    R.string.dialog_contrib_usage_limit_accounts,
                    FREE_ACCOUNTS,
                    currentLicence
                )

                PLANS_UNLIMITED -> context.getString(
                    R.string.dialog_contrib_usage_limit_plans,
                    FREE_PLANS,
                    currentLicence
                )

                SPLIT_TEMPLATE -> context.getString(
                    R.string.dialog_contrib_usage_limit_split_templates,
                    currentLicence
                )

                else -> null
            }

            TrialMode.DURATION ->
                if (licenceHandler.getEndOfTrial(this) < System.currentTimeMillis())
                    getLimitReachedWarning(context) else HtmlCompat.fromHtml(
                    buildString {
                        append(context.getString(
                            R.string.dialog_contrib_trial_info,
                            currentLicence,
                            "<i>" + context.getString(labelResId) + "</i>",
                            TRIAL_DURATION_DAYS,
                            Utils.getDateFormatSafe(context)
                                .format(Date(licenceHandler.getEndOfTrial(this@ContribFeature)))
                        ))
                        if (isPlay) {
                            append(" ")
                            append(context.getString(R.string.dialog_contrib_trial_info_play))
                        }
                    },
                    HtmlCompat.FROM_HTML_MODE_LEGACY
                )

            TrialMode.UNLIMITED -> context.getString(
                R.string.dialog_contrib_usage_limit_with_dialog,
                currentLicence
            )
        }
    }

    protected fun getCurrentLicence(context: Context, licenceHandler: LicenceHandler) =
        context.getString(licenceHandler.licenceStatus?.resId ?: R.string.licence_status_free)

    fun buildRemoveLimitation(ctx: Context, asHTML: Boolean): CharSequence {
        val resId = R.string.dialog_contrib_reminder_remove_limitation
        return if (asHTML) Utils.getTextWithAppName(ctx, resId) else ctx.getString(resId).replace(
            String.format("{%s}", Utils.PLACEHOLDER_APP_NAME), ctx.getString(R.string.app_name)
        )
    }

    val isExtended: Boolean
        get() = licenceStatus === LicenceStatus.EXTENDED
    val isProfessional: Boolean
        get() = licenceStatus === LicenceStatus.PROFESSIONAL

    companion object {
        const val FREE_PLANS = 3
        const val FREE_ACCOUNTS = 5
        const val FREE_SPLIT_TEMPLATES = 1
    }
}